src/app/clusters/webrtc-transport-provider-server/webrtc-transport-provider-server.cpp

```diff
@@ -0,0 +1,501 @@
+    bool deferredOffer = false;
+
+    auto status = Protocols::InteractionModel::ClusterStatusCode(mDelegate->HandleSolicitOffer(args, outSession, deferredOffer));
+    if (!status.IsSuccess())
+    {
+        // Delegate encountered a problem (out of resources, invalid data, etc.)
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    // Store or update the WebRTCSessionTypeStruct in the CurrentSessions
+    if (UpsertSession(outSession) == UpsertResultEnum::kInserted)
+    {
+        ChipLogProgress(Zcl, "WebRTCTransportProvider: Inserted a new session with ID=%u.", outSession.id);
+    }
+    else
+    {
+        ChipLogProgress(Zcl, "WebRTCTransportProvider: Updated existing session with ID=%u.", outSession.id);
+    }
+
+    // Send the response
+    Commands::SolicitOfferResponse::Type resp;
+    resp.webRTCSessionID = outSession.id;
+    resp.deferredOffer   = deferredOffer;
+    resp.videoStreamID.SetValue(outSession.videoStreamID);
+    resp.audioStreamID.SetValue(outSession.audioStreamID);
+
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, resp);
+}
+
+void WebRTCTransportProviderServer::HandleProvideOffer(HandlerContext & ctx, const Commands::ProvideOffer::DecodableType & req)
+{
+    // Extract command fields from the request.
+    auto webRTCSessionID = req.webRTCSessionID;
+    auto videoStreamID   = req.videoStreamID;
+    auto audioStreamID   = req.audioStreamID;
+
+    NodeId peerNodeId           = GetNodeIdFromCtx(ctx.mCommandHandler);
+    FabricIndex peerFabricIndex = ctx.mCommandHandler.GetAccessingFabricIndex();
+
+    uint16_t sessionId = 0;
+
+    // Determine if this is a new session or a re-Offer
+    if (webRTCSessionID.IsNull())
+    {
+        // This is a request for a new session
+        // If both the VideoStreamID and AudioStreamID are not present
+        if (!videoStreamID.HasValue() && !audioStreamID.HasValue())
+        {
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+            return;
+        }
+
+        sessionId = GenerateSessionID();
+    }
+    else
+    {
+        // This is a re-offer. Check if the session ID is in our provider list
+        sessionId                                 = webRTCSessionID.Value();
+        WebRTCSessionTypeStruct * existingSession = FindSession(sessionId);
+
+        if (!existingSession)
+        {
+            ChipLogError(Zcl, "HandleProvideOffer: No existing session with ID=%u.", sessionId);
+
+            // If WebRTCSessionID is not null and does not match a value in current sessions.
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+            return;
+        }
+
+        // Also check that the existing session belongs to the same PeerNodeID / Fabric
+        // If it doesn’t match, respond with CONSTRAINT_ERROR
+        if (peerNodeId != existingSession->peerNodeID || peerFabricIndex != existingSession->peerFabricIndex)
```

Comment:
I don't see this check in the spec.

Reply:
"If the accessing Peer Node ID and Local Fabric Index entries stored in the Secure Session Context, does not equal the PeerNodeID and FabricIndex for the WebRTCSessionID entry in [CurrentSessions]"

Note, FabricIndex has been removed from WebRTCSessionStruct